### DISCUSSION TITLE
Homemade Linux cloud saves for GOG Games using Heroic

### DISCUSSION THREAD

**Discussion by sluedecke** [mtkennerly/ludusavi#192] [General]
**Not working as of 2022-10-26, see https://github.com/mtkennerly/ludusavi/issues/129#issuecomment-1292648904**

This is more of a writeup / howto and not a bug report.

### Motivation

I have 200+ games on GOG and I've been using Linux exclusively for many, many years  Since I now own a steam deck, backing up save games is just the first - but important - step towards a missing feature GOG does not provide yet: Linux cloud saves.  While there are approaches to use the GOG api for that, I prefer to use my own devices.  The approach I take is described below.

I tested this approach successfully with:
* Dead Cells
* Terraria

### Here be dragons

* **Use it at your own risk.  It will eat your save games, including the one you invested so much love and so many hours! So please make a backup first!**

* **This is work in progress**

* **There is no error handling yet**

* **Script below works with GOG games only**

* **if you save on pc and deck without synchronisation (= while offline), conflicts will arise**

#### Wrapper script `ludusavi-cloud.sh`

* [deck] save it to something like `/home/deck/bin/ludusavi-cloud.sh`
* [pc] save it to something like `/home/MYUSER/bin/ludusavi-cloud.sh`

#### Game Launcher which supports a wrapper script (e.g. Heroic)

Heroic is an excellent choice and the following will assume you use it, but any game launcher supporting a wrapper script is fine.

* [deck] switch to desktop mode, open Discover
* [deck] install heroic  
* [deck] configure a default installation path, e.g. `/home/deck/Games`
* [deck] install a GAME, run it once
* [deck] add  `/home/deck/bin/ludusavi-cloud.sh` as a wrapper to your game
* [pc] install heroic
* [pc] configure a default installation path, e.g. `/home/MYUSER/Games`
* [pc] also install GAME, run it once
* [pc] add  `/home/MYUSER/bin/ludusavi-cloud.sh` as a wrapper to your game

#### Ludusavi

* [deck] switch to desktop mode
* [deck] create a folder `/home/deck/ludusavi-backup`
* [deck] open Discover
* [deck] install `ludusavi`
* [deck] configure ludusavi like this:
  * select backup mode
    * set backup path to `/home/deck/ludusavi-backup`
    * I prefer zip format since it saves some space
    * enable merge, use at least 1 full and 3 differential (although this is personal preference)
    * run "preview" to check if your GAME is detected.  If not, run GAME and save at least once.  If ludusavi still does not detect your game, check ludusavi documentation.
   * select restore mode and add redirects. ludusavi saves absolute filenames which need to be mapped on restoration to your target system (order is important!):
     * Add a redirect: source: `/home/MYUSER/Games`, target: `/home/deck/Games`
     * Add another redirect: source: `/home/MYUSER`, target: `/home/deck`.
* [pc] install `ludusavi`
* [pc] configure ludusavi like above, but use `/home/MYUSER/ludusavi-backup` as backup path and for the redirects map source: `/home/deck/Games` to target: `/home/MYUSER/Games` and source: `/home/deck`, target: `/home/MYUSER`

#### TBD Synchronisation of backup folder (e.g. Syncthing)

* [syncthing](https://syncthing.net/), configured to share the backup folder and running as service on deck and pc. For replication to work, both need to be running or you need a third installation of syncthing on a server. I use my NAS for that since it is always on.

### The script

```shell
#!/bin/sh

# -- ABOUT --
#
# Wrapper script to use ludusavi for savegame backup / restoration.
#
# Assumes that $3 is the directory with the game install which contains the file
# `gameinfo`.
#
# This file comes without warranty, use at your own risk!
#
# License: MIT License
# Author: Sascha Lüdecke <sascha@meta-x.de>

# -- HISTORY --
#
# Version: 0.2 - 2022-09-29
#
# . use zenity to tell user that we back up / restore
#
# Version: 0.1 - 2022-09-27

SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
LOGFILE=$SCRIPT_DIR/ludusavi-cloud.log
GAMENAME=`head -1 "$3/gameinfo"`

# Standard paths
LUDUSAVI=/usr/bin/ludusavi
ZENITY=/usr/bin/zenity
TEE=/usr/bin/tee

# Path on steam deck if installed via flatpack
if [ $USER == deck ]
then
    LUDUSAVI=/home/.steamos/offload/var/lib/flatpak/exports/bin/com.github.mtkennerly.ludusavi
fi

{
    echo ==================================================
    echo
    echo Gamename is: $GAMENAME
    echo Start time: `date`

    # restore savegame
    echo $LUDUSAVI restore --force "$GAMENAME"
    (
        echo "# Restoring savegame for $GAMENAME"
        # bypass STDOUT and explicitely redirect output to LOGFILE
        $LUDUSAVI restore --force "$GAMENAME" 1>&2 | $TEE -a $LOGFILE
    ) | $ZENITY --progress \
               --title="Savegame restore" \
               --no-cancel \
               --auto-close \
               --pulsate

    # run game
    echo Game run command:
    echo $@
    "$@"

    # backup savegame
    echo $LUDUSAVI backup --merge --force "$GAMENAME"
    (
        echo "# Backing up savegames for $GAMENAME"
        # bypass STDOUT and explicitely redirect output to LOGFILE
        $LUDUSAVI backup --merge --force "$GAMENAME" 1>&2 | $TEE -a $LOGFILE
    ) | $ZENITY --progress \
               --title="Savegame backup" \
               --no-cancel \
               --auto-close \
               --pulsate

    echo End time: `date`
    echo
    echo ==================================================
}  2>&1 | $TEE -a $LOGFILE
```

**Comment by sluedecke** [comment]
So ... I almost have it working.  The shell script runs fine, synthing does it's job.  BUT:

Restoring does not work IF the game is installed in a different folder.  I tried Dead Cells which is stored here:

* Deck: `/run/media/mmcblk0p1/gog/Dead Cells` (SD card)
* PC: `/home/MYUSER/Games/Dead Cells`

Is there any easy way to handle this in ludusavi?

**Comment by mtkennerly** [comment]
I think backup redirects (#72) would make this easier to pull off.

Right now, you would need each system to store its own backup folder, set up inverse restoration redirects on each system, and then restore from whichever system's backup is newer. It should work, but it's not ideal. Since you're using a script, you could implement logic like this:

* `ludusavi backups --api --path .../deck-backup $GAME` and check the latest timestamp
* `ludusavi backups --api --path .../desktop-backup $GAME` and check the latest timestamp
* take the backup folder with the newest backup and run `ludusavi restore --path $PATH ... $GAME`

Once backup redirects are implemented, then both systems could share a backup folder (as long as they use the same OS or otherwise have saves that map nicely), so you wouldn't have to worry about which one has the latest backup for the game.

**Comment by sluedecke** [comment]
Redirects sound about right, if I understand correctly that it is a concept ludusavi could provide.

Sharing the folder itself can then be left to the user, either use syncthing or a cloud drive or whatever keeps the backup folder accessible on all machines.

Which changes would ludusavi need in order to have redirects?  Maybe some additions to the mapping.yaml file which separate the root folder from the install_dir.  And then a detection mechanism which is able to find the game in a different root during restoration?

**Comment by mtkennerly** [comment]
Restoration redirects are already available, so backup redirects would work the same way. No need to modify mapping.yaml; it just needs to generically replace text in a target file path ([code](https://github.com/mtkennerly/ludusavi/blob/6574401658ee9ca5a035c5ce498f1bef8ee4536b/src/prelude.rs#L315-L318)).

What you can do right now, using only restoration redirects:

* Each system has its own backup folder, and both backup folders are synced to both systems
* Deck has a restoration redirect from `/home/MYUSER/Games` to `/run/media/mmcblk0p1/gog`
* Desktop has a restoration redirect from `/run/media/mmcblk0p1/gog` to `/home/MYUSER/Games`
* Your script decides whether to restore from the Deck or desktop backup

What you could do in the future, with both kinds of redirects:

* Both systems share a single, synced backup folder
* Deck has:
  * a backup redirect from `/run/media/mmcblk0p1/gog` to `/gog`
  * a restoration redirect from `/gog` to `/run/media/mmcblk0p1/gog`
* Desktop has:
  * a backup redirect from `/home/MYUSER/Games` to `/gog`
  * a restoration redirect from `/gog` to `/home/MYUSER/Games`

**Comment by sluedecke** [comment]
Very nice, I obviously missed that feature and it does the trick!  

At least for ''Dead Cells" it works with a single, synchronized backup folder.  Config excerpts:

On deck:
```yaml
backup:
  path: /home/deck/ludusavi-backup
restore:
  path: /home/deck/ludusavi-backup
  ignoredGames: []
  recentGames: []
  redirects:
    - source: /home/MYUSER/Games
      target: /run/media/mmcblk0p1/gog
```

On PC:
```yaml
backup:
  path: ~/Games/00-savegames/0-ludusavi-backup
restore:
  path: ~/Games/00-savegames/0-ludusavi-backup
  redirects:
    - source: /run/media/mmcblk0p1/gog
      target: /home/MYUSER/Games
```

Still .... this is not very easy to set up, because I need to add a mapping for each root.  This includes %APPDATA%, $HOME, ...  Also it works for a 2 device setting.  Having three devices with three different root requires much more mappings.

I would highly appreciate having the redirects in both ways.  I'd love to help, but I am way to new to rust to figure it out in a time efficient manner. Can you give me some hints so that I can work on a patch?

**Comment by sluedecke** [comment]
BTW: restoring with a redirect complains about `fluent-no-message=cli-game-line-redirected-from`

**Comment by mtkennerly** [comment]
Whoops, thanks for finding that. I'll include that in 0.13.1 shortly.

> I would highly appreciate having the redirects in both ways. I'd love to help, but I am way to new to rust to figure it out in a time efficient manner. Can you give me some hints so that I can work on a patch?

No worries! It might be a bit more complicated since it will touch various parts of the app (e.g., displaying the original path, adding the redirect editor, ensuring relevant paths are wrapped in the handling). I'll see if I can work on this within the next week or so, but if not, then I'll try to write something up. Thank you for your interest in contributing :)

**Comment by mtkennerly** [comment]
@sluedecke If you're interested in trying it out, I've implemented #72 and updated master. The config format for redirects has changed (and they're on the "other" screen now), which is automatically adjusted when upgrading, but won't be reverted when downgrading, so you may want to back up your config first.

You can do something like this now:

### Desktop
> <img width="702" alt="image" src="https://user-images.githubusercontent.com/4934192/193447184-4d55805b-f418-4140-8c73-28796394df61.png">

### Deck
> <img width="698" alt="image" src="https://user-images.githubusercontent.com/4934192/193447196-482fb6d0-1129-470e-91b9-d12d9d163122.png">

**Comment by sluedecke** [comment]
I like the new bidirectional mappings, they work fine for me.  Thanks for adding them, it really helps keeping things tidy!

**Comment by sluedecke** [comment]
As of 2022-10-26 this is not working properly.  I tested it by accident with a non-flatpak installation of heroic.

Using heroic the flatpak and thus most common way does not allow one to run ludusavi from the wrapper script.  Apparently the flatpak runtime used in heroic is missing libbz2 which is needed by ludusavi.

**Comment by mtkennerly** [comment]
Do you have bzip2 compression enabled for backups? I wonder if it's failing immediately when Ludusavi starts or only when it tries to compress a backup.

**Comment by sluedecke** [comment]
It does not start, not even when using `--preview` on either backup or restore.

update: probably an issue for heroic to include this library in their flatpak runtime.  The game "Dead Cells" is missing libbsd for example: https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/issues/1829
